### ISSUE TITLE
[Question] WSAD

### ISSUE THREAD

**Issue by SilentMRG** [yquake2/yquake2#1022]
Hello...

Binding WSAD just doesn't work and I can't move the character. The arrows work, but it's horrible to use them. Is there any way to get WSAD working or is it something the engine can't do? In the GOG version, WSAD works without problems.

Thank you for your attention!

**Comment by protocultor** [comment]
You'll need to give more info about your setup.
Which Yamagi version? Which OS? I'd even ask for your keyboard and how do you connect it to your computer.
I'm asking because I've tried YQ2 in many things, including a Raspberry Pi, and in none of them something as basic as defining keyboard controls has failed. People have been using WASD for years, and YQ2 has no issues with it.

**Comment by SilentMRG** [comment]
Thanks for the quick response! But no additional information is required. I managed to fix the "problem". The "yq2.cfg" file, located in the "misc" folder, was having full privileges and preventing me from binding the keys, so I created an "autoexec.cfg" in the "baseq2" folder with the following content:

//
// KEY BINDINGS
//

unbindall

bind	1		"use Blaster"
bind	2		"use Shotgun"
bind	3		"use Super Shotgun"
bind	4		"use Machinegun"
bind	5		"use Chaingun"
bind	6		"use Grenade Launcher"
bind	7		"use Rocket Launcher"
bind	8		"use HyperBlaster"
bind	9		"use Railgun"
bind	0		"use BFG10K"

bind	MOUSE1		+attack

bind	s		+back
bind	w		+forward
bind	a		+moveleft
bind			+left
bind	d		+moveright
bind			+right

bind	TAB		inven
bind	ENTER		invuse
bind	z		invprev
bind	x		invnext
bind	BACKSPACE	invdrop

bind	MWHEELDOWN	weapprev
bind	MWHEELUP	weapnext

bind	q		"use quad damage"
bind	e		"use power shield"

bind	SHIFT		+speed

bind	c		+movedown
bind	SPACE		+moveup

bind	PAUSE		pause

bind	F1		"cmd help"
bind	F5		"echo Quick Saving...; wait; save quick"
bind	F9		"echo Quick Loading...; wait; load quick"
bind	F12		screenshot jpg 90

bind	t		messagemode

//----------------------------------------------

And now everything is working like a charm! Just a note: In the documentation it is said to move the "yq2.cfg" file to the "baseq2" folder, for it to work, actually moving it is not necessary, as it works directly from the "misc" folder, this in version 8.20.

**Comment by protocultor** [comment]
Good that you could solve your problem... but that's not how `yq2.cfg` works.
That file is completely optional, and exists just to be able to apply default controls without having to replace `default.cfg` inside the pak file(s).
Also, it's not normally possible for it to be read outside the "current mod" directory (`baseq2` for the base game), so I don't think that was the origin of your problems.
I'm assuming you are using Windows, so, can you see if you have a `My Documents\YamagiQ2\baseq2` folder? Do you have a `config.cfg` file there? That's where your video and controls configurations are saved (unless you are using the `-portable` option, which is unreliable and not recommended), and your config should always override anything in `yq2.cfg`, `default.cfg`, and even the autoexec.

**Comment by SilentMRG** [comment]
You are right, there is a file in "my documents". According to the documentation, it is said that the Windows version is fully portable and that it is possible to move the folder wherever you want. So at no point did I think there might be a folder under "my documents". Is this "-portable" parameter to be placed in the game's executable or is it to be created a .txt file with this parameter? Just to know. I like portable source ports better.

And again, thanks for the information!

**Comment by protocultor** [comment]
I asked you for the existence of the `YamagiQ2` folder because it didn't seem your config was being saved.
If you enter `customize controls` in Options, and change your keys there, nothing else can override what you set until you delete your own `config.cfg`.
But if you have your video options saved, the config works as intended.

In the documentation, by "portable" they mean you don't need Q2 to be in a specific directory; you can put it anywhere in your hard drive. Config files and savegames are another history, and depend on the user currently logged in (that's why they end up in `My Documents`).

The `-portable` parameter is added when you run the YQ2 executable ([in the docs](https://github.com/yquake2/yquake2/blob/master/doc/040_cvarlist.md#command-line-arguments)). The shortest way is doing a shortcut to `yquake2.exe`, go to Properties and add ` -portable` in the Target, at the end. Again, this is unreliable, so you might get a new set of problems for using it (or not).

**Comment by SilentMRG** [comment]
> I asked you for the existence of the `YamagiQ2` folder because it didn't seem your config was being saved.
> If you enter `customize controls` in Options, and change your keys there, nothing else can override what you set until you delete your own `config.cfg`.
> But if you have your video options saved, the config works as intended.

This is the problem: I changed the keys in the game options and nothing was changed in fact, there in the options, the keys were bind, but when it was time to play, the keys were not working, only the arrows; and that's when I I had to take a closer look at the files to see what was going on...

> In the documentation, by "portable" they mean you don't need Q2 to be in a specific directory; you can put it anywhere in your hard drive. Config files and savegames are another history, and depend on the user currently logged in (that's why they end up in `My Documents`).

Let me get this straight... So as long as I'm logged into my account, I can move the folder wherever I want?

> The `-portable` parameter is added when you run the YQ2 executable ([in the docs](https://github.com/yquake2/yquake2/blob/master/doc/040_cvarlist.md#command-line-arguments)). The shortest way is doing a shortcut to `yquake2.exe`, go to Properties and add ` -portable` in the Target, at the end. Again, this is unreliable, so you might get a new set of problems for using it (or not).

Well, if it's not reliable, then better leave it the way it is. yquake2 is reminding me of DevilutionX (source port of Diablo) as it works the same way regarding game save files and other settings.

Thank you very much in advance for all the clarifications!

**Comment by SilentMRG** [comment]
Forgot to mention that moving the yq2.cfg file to the baseq2 folder didn't help either. The keys were remapped in the game's options menu, but when playing, they worked according to yq2.cfg bindings, in other words, yq2.cfg ignored my bindings. Then I created the autoexec.cfg file and everything worked fine.

**Comment by protocultor** [comment]
When you are using your Windows computer, you are logged in. Windows is a "multi-user" OS, and when you install it, it forces you to create at least a single user, which you use to operate the computer.

Then, anything that could change depending on the user, is saved to a folder identified by something similar to the _XDG Base Directory Specification_ ([read its basics](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html#introduction)). In the case of YQ2, this is your user's `My Documents` folder.

As your config and savegames in YQ2 depend on the current Windows user, you can't change the location of the config (or at least you can't change it and expect it to keep working) unless you use the `-portable` parameter. Which you should try anyway, because what happens in your machine is very strange, doesn't makes sense to me, and I don't think you could make it worse. At worst, you can get rid of the parameter to get back to your current situation.

**Comment by SilentMRG** [comment]
Thanks for the info...
Just for the record: I'm using a Laptop with Windows 7 Ultimate 64-bit and an ABNT2 (Brazilian Portuguese) keyboard. Well, I moved the folder from drive C to drive D (without deleting the folder in "my documents") and the only problem I got was a random APPCRASH error when I quit the game. All other things are working fine, including the other games I have here. Perhaps, what I had was a one-off source port bug, in other words, I was the "lucky one chosen".
